README.md
letownbuilder
=============

A Town building simulator

Ever played an adictive online game where you build your kingdom or town and realized that it was taking too long to do anything unless you spent real money?   LeTownbuilder allows you to scrach that "building and managing" itch while not burdening you with microtransations, or massive hardware requirements.

Your town begins in a vacant, unsettled area.  Your start location is randomized and there are different possible biomes, so no 2 times you play are the same.  You start with 5 villagers.  Start by collecting resources and building houses and specialized buildings.  Villagers have moods, and can become happy or angry based many factors. Keep them fed, and housed.  Also make sure that there is a bonfire at night, or they will get cold.  You can increase happiness by always having food at lunchtime and by adding other advanced upgrades to your village(TBD)


Current villager types
============================
* normal,peasant - Your average villager comes in male female, or child verieties.  They are the average hard working villager with no bonus or penalty.

* child - The future of the settlement relies on these youngsters.  They dont work, and are an extra mouth to feed.

* bandits - Bandits are a mischivious breed who will attempt to steal food at lunchtime and other items(tbd)  They currently have no advantage but may later have advantages in hunting or assasinating unloved mayors.

* holymen - These preachers are not "affected" by general village unrest.  They don't particularly like dark elves.  They're firey sermons can provide a daily bonus to the general village morale.  However if the village morale drops too low, their sermons may have the opposite effect.

* werewolves - werewolves look like regular citizens until the sun goes down.  They then become vicious hunters and can kill villagers!  there is currenly no solution to handling werewolves, but (tbd) sheriff's will soon be able to deal with them.  Dark elves will also be able to handle them in a limited scope

* dark elves - the often misunderstood dark elf provides a massive bonus to research.  Having a dark elf, means free research without a school, and also allows you to research at a faster speed. However(tbd) dark elves are pyromaniacs and will randomly start fires.  They are also (TBD) the chief suspect besides the bandit who is responsible for the mayors untimely death.

* ghosts - ghosts are the departed who may be seen at night, if not burried in a graveyard...

Current villager roles
============================
* Mayor - Mayors are the source to blame for all things aray in your village.  Often villagers will proclaim themselves to be mayor at the founding of your village.  If this does not occour, you can later elect one after researching civics.
* Farmer - farmers are responsible for planting and caring for farm plots
* Builder - Builders build.
* Sheriff - The sheriff deals with riff/raff.
* Fisherman - Professional food gatherer if there is a body of water nearby.
* (tbd) Firefighter - Will provide a way to fight wildfires/spreading fires...

Research types
===============================
Agriculture - the upgrading of farms and providing various new farms.
* Level 1 Garden Fencing.  Decrease the chance of crows stealing crops.
* Level 2 (tbd) Tomato farms. Adding food variety increases villager happiness.
* Level 3 (tbd) Chicken farms.  Add pultry to your diet.
* Level 4 (tbd) Cooking.  Cook your food to reduce the risk of food contamination. 
* Level 5 (tbd) Treefarms. Get fruit.
* Level 6 (tbd) Fishfarms. Make use out of all those holes you dug.

Economy - the establishment of tarde and markets
* Level 1 Trade Post. (Tbd) will be able to accept merchants.
* Level 2 (tbd) Traders. Can send out traders to other towns.
* Level 3 (tbd) Riverboating (tbd - if no water just better prices)
* Level 4 (tbd) Trade treaties.  You can now garentee goods and services for better prices.
* Level 5 (tbd) Currency. You can now buy and sell with coin.
* Level 6 (tbd) Banks. Reposses your homes, knowing that your economy will improve in the long run?
* Level 7 (tbd) Shipping. Mass moving of product.
* Level 8 (tbd)

Security - With a growing number of dangers being added to the game, security will play a larger role in keeping your town safe
* Level 1 Sheriff's Office/sheriff You can now catch food thieves.
* Level 2 (tbd) militia. The Sheriff will be able to call on the militia for emergancies.
* Level 3 (tbd) hunters. cull the local wildlife and get food for doing it.
* Level 4 (tbd) Palasade - build walls around your town.
* Level 5 (tbd) Deputies - Reduce the need for militia with full time sub-sheriffs.
* Level 6 (tbd) Rangers - Bane of werewolves, and bandits.
* Level 7 (tbd) Punishments - Make an example of evildoers.(decrease unrest)

Industry - Industry allows you to build more advanced building which produce more.
* Level 1 fishermans hut - Now provides a fresh catch daily without manual fishing.
* Level 2 (tbd) inns - inns/innkeepers provide extra town stability by allowing tourism and family/friends.
* Level 3 (tbd) hospital - doctors heal the sick...sometimes.
* Level 4 (tbd) Improved Mining. You now can mine metals and precious metals.
* Level 5 (tbd) Workshops.  Make things to export!
* Level 6 (tbd) Alchemist.  Make medicinces, perfumes and poisons!
* Level 7 (tbd) Smelting. Make metal!

Civics - Civics will allow you to manage a town more effectivly by making the town more stable.  Things like town halls, council, city planning, and more are planned for the civics line of research.
* Level 1 (tbd) Mayor's monument. A symbol of the mayers grandure, or pomposity? It ups the happiness of the town, yet is the first to get defaced if things go bad.
* Level 2 (tbd) Town Hall.
* Level 3 (tbd) Free elections.  Finally choose your own mayor!
* Level 4 (tbd) Town council.  Someone else makes the smaller decisions and gives the mayor culpible denyability.
* Level 5 (tbd) Propaganda! When the going gets bad, the bad start lying...
* Level 6 (tbd) Palace.  The mayor gets his own house finally!
* Level 7 (tbd) Upper class. Nobles; another annoying person to bow to.
	
Research Combos (tbd)
================================
* Agriculture(1), Security(1) (tbd) Scarecrow - Further reduce bird raids.
* Civics(2) Security(2) Militia House, a permanent place for your citizen army, reduces unrest and allows for a longer militia muster time.
* Industry(1) Economy(3) Fisherboats! Catch more fish out in the middle of a lake.
* Agriculture(4) Economy(5) Restaurants. Eat out in style.  Comfort food for massive unrest reduction.
* Agriculture(1) Industry(1) Civics(1) Waterforming.  Fill holes with water.

Wildlife types (tbd) wildlife will provide wild animales which can be hunted or, in some cases WILL HUNT YOU!
================================
rabbit - little vermin that taste good if cooked properly(tbd)  (tbd rabid rabits) can bite villagers and give you rabbies!
wild donkey - pack animals, that dont pack because they are wild.(tbd animal training?)

Basic building types
================================
* House - houses are needed for providing villagers with homes.  Only one house is needed per family
* Barn -  barns store food.  They keep from having too many food icons on the screen.  More importantly they provide for long term food storage.  Food tends to rot after a while without a barn.
* Mine - Mines provide a daily allotment of stone without digging holes all over the landscape.  (tbd) they will also later provide metals and precious metals.
* School - Schools allow a town to start reseaching.  (tbd) they will also later provide children with something to do during the daytime.

Advanced building types
================================
* Trade post - trade posts will allow merchants to trade in your town (TBD merchants)
* Sheriff's Office - A sheriff's office will allow you to(TBD) assign a sheriff who will be responsible for town security.  Sheriff's have a chance of catching bandits stealing food, dark elves starting fires and werewolf rampages.  The chance of preventing these catastrophes increases with each level of security.
* Fisherman's hut - a fishermans hut will assign a villager to be a dedicated fisherman.  The hut must be built near water and provides a number of fish per day.
* TODO: Drawn/not coded, militia house
* TODO: Drawn/not coded, smelter - smelted.
FAQ's
=================================
Q. How do I get letownbuilder working on my andriod?
A. There are a few things that you need to do to get letwon builder installed on your tablet.
  1. Download the love apk at 
  2. love2d source port to android - love-native-android
     Port of the Love2D source to android using ndk.

      Repository: 
      https://github.com/hagish/love-native-android

      Download:
      http://www.brightidea.eu/games/love-native-android.apk 
  3. Install it (your andriod must be able to install non market apps)
  4. Get a copy of the code, zip it up, and name it a letownbuilder.love
  5. play it through the app!
  6. (not currently there is a scrolling bug in the andriod version?

Q. Feature X appears to be missing.
A. If you have any ideas please feel free to email me at jobs@cooperlabs.net

Q. I found a lua script error!
A. Eeek!  PLEASE email me imediatly at jobs@cooperlabs.net, include error and line number.





Changelog.txt
Readme
=================================

Villagers should go to the barn at noon
Kids should go to school at 8 (and it ends when they go to the barn)

Planned features

BUGS: Message "Fordic has fought a werewolf and lived!" (Fordic is dead, and the records state that he defeated a nightwolf in single combat)
BUGS: merchants keep moving even if paused.
BUGS: Cannot build a smithy, select house to build, choose smithy, select where, and nothing happens.
BUGS: Scarecrows dont show up.

Deserts
 DONE: Gathering food in green tiles yields ?what?, and tiles turn back to sand.
 DONE: Cutting trees in the tree tiles should yield wood.
 DONE: Digging a hole should yield sand/sandstone.
 DONE: Fishing in water should yield (fewer) fish.
 TODO: (bug?) Should gardens work in sand?
 DONE: Houses/mines should cost 1 wood, 6 sandstone
 TODO: 

issue: display dead villagers as red in records
DONE: (Fixed)carrots can be collected by mining in deserts.
DONE: (Fixed)messages button does not highlight in records, but food highlights it.
bugs: (verified) grain shows up as tomatoes if tomatoes is researched.
bugs: (verified) grain at stage one shows up as tomatoes fully grown.
DONE: (to verify)merchants.lua 255 (use of game_sprites) -added draw_small_sprite
DONE: nightwolves do not kill
bigs: too many werewolves(lowered troublemaker chance)

BUGS: currently smelter and smithy are confused in some situations (like starting dwarves)
fixed: House count no longer updates/stays updated. [FIXED]
fixed?: fire spreads where no trees are, cant build trade post even with enough wood
fixed?: wood shows -15(minus) in japanese biome (something subtracting wood not sakura)
fixed?: forward slash roads do not connect. Needs pixel correct.
fixed?: Too much talking! decrease by increasing the chance of text by random (1, table.getn(game_villagers)*100)
fixed?: zombie shows when animal is bagged in hunting.
fixed?: you can hunt already killed animals (and they can kill you!)
fixed: message log flows offscreen with prolonged play. - [FIXED]
fixed: Sharif's office shows vandalized monument, and claims debug "barn"
fixed: tradepost shows up as shariff's office
DONE: graveyard shows "debug tried to build a barn" but builds.
fixed: check fireproof now should work.
DONE: working gardens (not giving food?)


TODOS: 


Weather
 DONE: Snow - create ice flakes falling down, requires at least 1 fire per 5 villagers to avoid freezing.[DONE]
 issue: create weather for snow and desert levels.
 * Snow - create snow on the ground, tiles get whiter...
 * Rain - causes  1/100th chance of causing flooding, 1/50th chance of putting out fires.
 * Rain - Flood trenches
 * Draught - lack of rain for 14+ days, causes a loss in plant life.  Green areas become dirt. Dirt does not regenerate until it rains. Drought has a 1/50 chance of ending.
 * Locusts! - all of your food are belong to us!  Locust has a 1/1000 chance of happening, and each green tile has a 9 /10 chance of being eaten.  All gardens reset to 0. Barns are untouched.  Any food not in a barn is eaten.
  
Bandit Raids - TBD
  * 1/500th chance of a bandit raid per day.
  * Bandits appear and steal between 75-95% of food.  
  * Resistance can cut the theft of food down to 25%, but risks loosing villagers. Between 0-half of villagers can die.  Sherriff and militia house can increase the chances of lower casualties.
  * Bandits can take other plunder like gold and captives.  Loosing villagers to captivity cause a massive drop in morale, and may cause loved ones to leave settlement to go look for them.
  * Bandit villagers may join in the looting. (25% chance)

Slaver Raids - TBD
  * Slavers appear and take 80% of population under the age of 45.
  * Resistance can be formed similar to bandit resistance.
  * Huge drop in morale.



4/25/2015
* rain fills full screen
* fixed loading villager duplication
* added dwarves, enables smithy and smelter if one migrates.
* added disabled menu item for desert biome
* added day and night music
* added donkey cart at "center" of town?
* message box shows up at bottom of the screen even in full screen

BIG Todos
* make roads buildable again
* mushrooms, saltwort, sandstone not showing up on resources icons.

4/1/2015  
*Tile 65, watchtower (TODO) tower provides light at night up to 2 squares.(TODO) Make watchtower a 
 security upgrade 1 and shift everything else up by one.
* moved to sprite atlas to fix memory problems
* animal propagation
* job queue, but still using debug
* added earthquake catastrophe (1/1000 per day)
* added zombie appacalypse? untested, (1/1000 per day)


12/21/2014 Added weather, and rioting caused fires.

12/8/2014 Added more research topics. (no art yet)  can now research melting, smithing, tools and weapons.

1/21/2014
* message que finally works, plays messages one after the other
* added diseases! villagers can get a disease from eating contaminated food!  This will provide an incentive to use farms, and cooking later on.
* other possibilities could be rabid rabbits, etc!

1/16/2014
* split main.lua file to make it more manageable.
* Added male dark elves
* Dark Elf children now appears as children (todo should not be a scion of a normal)
* Added Sheriff sprite, sheriffs talk.
* Graveyards now show in both biomes correctly(todo, make a max bodies per gy
* Ghosts now appear, move around, but dont talk...why?
* Message log now shows up if you press r (with date/time of message)
* fixed house save/load bug with game_road_map table
* Villagers tasked wtih a job will move toward job area
* Fishing huts now produce fish!
* Villagers talk for less time.

1/14/2014
* Building animations work now
* (previous) now there is a roster of villagers
=================================
1/1/2014
* Added fullscreen - can set from a variable on startup or can press F2
* Fixed a few bugs with fullscreen, buttons now resize with screen
* Fixed bugs with bridges showing up as a house

animals.lua
-- file for animals, wildlife and pets.
-- test edit
function new_villager_pet(pet_type, pet_owner)
   local wildlife_types = {"bunny", "donkey",
			   "llama", "sheep", 
			   "shittzu dog", "pitbull dog", "mongrel dog", "dalmation dog",
			   "tabby cat", "black cat", "ferret",
			   "cow", "bull","parrot"
   } --added donkey art...
   a = {x = start_x, y = start_y, dx = start_x, dy = start_y, 
	died_x = 0, died_y = 0, direction = 0, job = 0, speak = 0, 
	talk_timer = 0, alive=1, --always start alive 
	domesticated = 0,
	sex = math.random(0,1), --1 male, 0 female
	age = 1, --TBD pet ages.
	wildlife_type = pet_type,
	owner = pet_owner,
	speed = 0}
   if a.wildlife_type == "bunny" then a.speed = 15 a.sprite = 17
   elseif a.wildlife_type == "donkey" then	a.speed = 3	a.sprite = 23
   elseif a.wildlife_type == "llama" then  a.speed = 3 a.sprite = 30
   elseif a.wildlife_type == "sheep" then a.speed = 3 a.sprite = 28
   elseif a.wildlife_type == "shittzu dog" then a.speed = 3 a.sprite = 31
   elseif a.wildlife_type == "pitbull dog" then a.speed = 3 a.sprite = 32
   elseif a.wildlife_type == "mongrel dog" then a.speed = 3 a.sprite = 32
   elseif a.wildlife_type == "dalmation dog" then a.speed = 3 a.sprite = 32
   elseif a.wildlife_type == "tabby cat" then a.speed = 4 a.sprite = 35
   elseif a.wildlife_type == "black cat" then a.speed = 4 a.sprite = 35
   elseif a.wildlife_type == "ferret" then a.speed = 4 a.sprite = 35
   elseif a.wildlife_type == "cow" then a.speed = 4 a.sprite = 35
   elseif a.wildlife_type == "bull" then a.speed = 4 a.sprite = 35
   elseif a.wildlife_type == "parrot" then a.speed = 4 a.sprite = 17
   elseif a.wildlife_type == "goblin" then a.speed = 3 a.sprite = 43 a.dead_sprite = 44
   elseif a.wildlife_type == "orc" then a.sprite = 45 a.dead_sprite = 46
   elseif a.wildlife_type == "skeleton" then a.sprite = 47 a.dead_sprite = 48
   end
   return a
end

function new_wildlife(migration, name)
   local start_x = math.random(100,900)
   local start_y = math.random(100,900)
   local randtype = math.random(1,2)
   local wildlife_types = {"rabbit", "wild donkey",
			   "wild llama", "wild sheep", "fox", 
			   "wild dog", "bear", "black bear",
			   "lynx","copperhead"} --added donkey art...
   local randtype = math.random(1,table.getn(wildlife_types) )
   
   a = {x = start_x, y = start_y, dx = start_x, dy = start_y, 
	died_x = 0, died_y = 0, direction = 0, job = 0, speak = 0, 
	talk_timer = 0, alive=1, --always start alive 
	domesticated = 0,
	sex = math.random(0,1), --1 male, 0 female
	age = nil, --math.random 18,65
	wildlife_type = wildlife_types[randtype],-- 1, normal, 2, dark elf, 3 bandit, 4 light elf
	speed = 0
   }
   if name == "random" then
      --nothing
   else
      a.wildlife_type = name
   end
   if migration == 0 then
      a.age = math.random(18,50)
      --a.wildlife_type = "rabbit"
   else
      a.age = math.random(1,70)
   end --if migration == 0 then
   if a.wildlife_type == "rabbit" then a.speed = 15 a.sprite = 17 a.dead_sprite = 18
   elseif a.wildlife_type == "wild donkey" then	a.speed = 3 a.sprite = 23 a.dead_sprite = 29
   elseif a.wildlife_type == "wild llama" then  a.speed = 3 a.sprite = 30 a.dead_sprite = 29
   elseif a.wildlife_type == "wild sheep" then a.speed = 3 a.sprite = 28 a.dead_sprite = 29
   elseif a.wildlife_type == "fox" then a.speed = 3 a.sprite = 31 a.dead_sprite = 29
   elseif a.wildlife_type == "wild dog" then a.speed = 3 a.sprite = 32 a.dead_sprite = 29
   elseif a.wildlife_type == "bear" then a.speed = 3 a.sprite = 33 a.dead_sprite = 29
   elseif a.wildlife_type == "black bear" then a.speed = 3 a.sprite = 34 a.dead_sprite = 29
   elseif a.wildlife_type == "lynx" then a.speed = 4 a.sprite = 35 a.dead_sprite = 29
   elseif a.wildlife_type == "copperhead" then a.speed = 4 a.sprite = 36 a.dead_sprite = 29
   elseif a.wildlife_type == "donkey" then a.speed = 4 a.sprite = 24 a.dead_sprite = 29
   elseif a.wildlife_type == "night wolf" then a.speed = 11 a.sprite = 42 a.dead_sprite = 29
   end
   return a
end

--no spawn all invaders.
function spawn_nightwolves(invader_type) --put night wolves into game_nightwolves = {}
   local nesw = math.random(1,4)
   --select a random tile on the edge of the map to place them
   --note code, try spawn offscreen.
   nesw = 1
   if nesw == 1 then --north
      wolf_spawn_x = 400 --math.random(1,game.tilecount) --tiles
      wolf_spawn_y = -400 --math.random(1,game.tilecount) --tiles
   elseif nesw == 2 then --east
      wolf_spawn_x = 600 --game.tilecount
      wolf_spawn_y = 600 --math.random(1,game.tilecount)
   elseif newsw == 3 then --south
      wolf_spawn_x = -600 --math.random(1,game.tilecount) --tiles
      wolf_spawn_y = 600 --game.tilecount --math.random(1,game.tilecount) --tiles
   elseif newsw == 4 then --west
      wolf_spawn_x = -600 --0 --tiles
      wolf_spawn_y = -600 --math.random(1,game.tilecount) --math.random(1,game.tilecount) --tiles
   end
   
   table.insert(game_nightwolves, new_wildlife(0, invader_type) ) --need to set x,y based on screen edge.
   table.insert(game_nightwolves, new_wildlife(0, invader_type) )
   table.insert(game_nightwolves, new_wildlife(0, invader_type) )
   table.insert(game_nightwolves, new_wildlife(0, invader_type) )
   table.insert(game_nightwolves, new_wildlife(0, invader_type) )
   
   for i,v in ipairs(game_nightwolves) do
      game_nightwolves[i].x = wolf_spawn_x+math.random(1,64)
      game_nightwolves[i].y = wolf_spawn_y+math.random(1,64)
   end
   --game_nightwolves = {}
end
function drop_nightwolves() --nights over wolves are gone.
   for i,v in ipairs(game_nightwolves) do
      table.remove(game_nightwolves)
   end
end

function update_nightwolves(dt)
   for i, v in ipairs(game_nightwolves) do
      if game_nightwolves[i].job == 0 and game_nightwolves[i].alive == 1 then -- The living!
	 game_nightwolves[i].job = math.random(1, 1000)
	 if game_nightwolves[i].job < 10 then -- movment
	    game_nightwolves[i].dx = game_nightwolves[i].dx + math.random(-30,30) --temp workaround
	    game_nightwolves[i].dy = game_nightwolves[i].dy + math.random(-30,30)
	 end--endif
      end --endif
      update_villager_new_destination(game_nightwolves[i], dt, game_nightwolves[i].speed)
   end
end

function draw_nightwolves()
   for i,v in ipairs(game_nightwolves) do
      draw_small_sprite(game_nightwolves[i].sprite,game_nightwolves[i].x +game.draw_x, game_nightwolves[i].y +game.draw_y )
   end
end

function wildlife_proliferation() --disabled because of memory bug
   --loop through current wildlife, and add wildlife based on type. some wildlife breed faster than others.
   for i,v in ipairs(game_wildlife) do
      local pop = math.random(1,5)
      local name = "none"
      if game_wildlife[i].wildlife_type == "rabbit" then
	 for t = 1, pop do
	    table.insert(game_wildlife,  new_wildlife(0, "rabbit") )
	 end
      elseif game_wildlife[i].wildlife_type == "wild donkey" or
	 game_wildlife[i].wildlife_type == "wild llama" or
	 game_wildlife[i].wildlife_type == "wild sheep" then
	 name = game_wildlife[i].wildlife_type
	   if pop == 1 then
	      table.insert(game_wildlife,  new_wildlife(0, name) )
	   end
      elseif game_wildlife[i].wildlife_type == "fox" or game_wildlife[i].wildlife_type == "wild dog" then
	 pop = math.random(1,3)
	 if pop == 1 then
	    name = game_wildlife[i].wildlife_type
	    table.insert(game_wildlife,  new_wildlife(0, name) )
	 end
      elseif game_wildlife[i].wildlife_type == "bear" or game_wildlife[i].wildlife_type == "black bear" then
	 pop = math.random(1,20)
	 if pop == 1 then
	    name = game_wildlife[i].wildlife_type
	    table.insert(game_wildlife,  new_wildlife(0, name) )
	 end
      elseif game_wildlife[i].wildlife_type == "lynx" or game_wildlife[i].wildlife_type == "copperhead" then
	 pop = math.random(1,5)
	 if pop == 1 then
	    name = game_wildlife[i].wildlife_type
	    table.insert(game_wildlife,  new_wildlife(0, name) )
	 end
      end
   end
   --notes, bears break into store houses with fish
   --coperhead's can bite fishermen! (1,1000 chance to attack a fisherman)
   --more than  7 rabbits can devistate crops!
   --wild dogs have (1,1000) chance of biting children
end
